### PR TITLE
Fix: re-registering SA with SAM raises error (#2407)

### PR DESCRIPTION
**Pull request by noah-aviel-dove**
Author (check before every review)

- [x] PR title references issue
- [x] Title of main commit references issue
- [x] PR is linked to Zenhub issue
- [x] Added `DCP/1` label                                   <sub>or this PR does not target an `hca/*` branch</sub>
- [x] Created issue to track porting these changes          <sub>or this PR does not need porting</sub> 
- [x] Added `r` tag to commit title                         <sub>or this PR does not require reindexing</sub>
- [x] Added `reindex` label to PR                           <sub>or this PR does not require reindexing</sub>
- [x] Freebies are blocked on this PR                       <sub>or there are no freebies in this PR</sub>
- [x] Freebies are referenced in commit titles              <sub>or there are no freebies in this PR</sub>
- [x] This PR is blocked by previous PR in chain            <sub>or this PR is not chained to another PR</sub>
- [x] Added `chain` label to the blocking PR                <sub>or this PR is not chained to another PR</sub>
- [x] Ran `make requirements_update`                        <sub>or this PR leaves requirements*.txt, common.mk and Makefile untouched</sub>
- [x] Added `R` tag to commit title                         <sub>or this PR leaves requirements*.txt untouched</sub>
- [x] Added `reqs` label to PR                              <sub>or this PR leaves requirements*.txt untouched</sub>
- [x] Documented upgrading of deployments in UPGRADING.rst  <sub>or this PR does not require upgrading</sub>
- [x] Added `u` tag to commit title                         <sub>or this PR does not require upgrading</sub>
- [x] Added `upgrade` label to PR                           <sub>or this PR does not require upgrading</sub>
- [x] `make integration_test` passes in personal deployment <sub>or this PR does not touch functionality that could break the IT</sub>
- [x] Rebased branch on `develop`

Primary reviewer (before pushing merge commit)

- [x] Checked `reindex` label and `r` commit title tag
- [x] Rebased and squashed branch
- [x] Sanity-checked history
- [x] Build passes in `sandbox`                             <sub>or added `no sandbox` label</sub>
- [x] Reindexed `sandbox`                                   <sub>or this PR does not require reindexing</sub>
- [x] Added PR reference to merge commit
- [x] Moved linked issue to Merged
- [ ] Pushed merge commit to Github

Primary reviewer (after pushing merge commit)

- [ ] Moved freebies to Merged column                       <sub>or there are no freebies in this PR</sub> 
- [ ] Shortened chain                                       <sub>or this PR is not the base of another PR</sub>
- [ ] Verified that `N reviews` labelling is accurate
- [ ] Commented on demo expectations                        <sub>or labeled as `no demo`</sub>
- [ ] Pushed merge commit to Gitlab                         <sub>or this changes can be pushed later, together with another PR</sub>
- [ ] Reindexed `dev`                                       <sub>or this PR does not require reindexing</sub>
- [ ] Deleted PR branch from Github and Gitlab
- [ ] Unassign reviewer from PR

### CODE REVIEW

**Comment by noah-aviel-dove** [comment]
To be clear, `make sam` did not _succeed_ before this PR, but it was not clear that the failure occurred during SAM registration.

File: src/azul/terra.py [hunk 1]
```diff
@@ -150,9 +150,10 @@ def register_with_sam(self) -> None:
             log.info('Google service account previously registered with SAM.')
         elif (response.status == 500
               and 'Cannot update googleSubjectId' in response.data.decode()):
-            log.warning('Unable to register %s. SAM does not allow re-registration of '
-                        'service account emails. Refer to the troubleshooting section '
-                        'of the README.', self.credentials.service_account_email)
+            raise RequirementError('Unable to register %s. SAM does not allow '
```
Comment (amarjandu) [inline src/azul/terra.py:153]: I like that the script is exited if this is hit, there appears to be a problem with the interpolation of the email into the text.

```
Apply complete! Resources: 3 added, 11 changed, 2 destroyed.
/Applications/Xcode.app/Contents/Developer/usr/bin/make link_health_checks
python /Users/amar/dev/hca/azul/scripts/manage_health_checks.py --link
2020-11-12 09:11:35,375 INFO    MainThread: Found credentials in environment variables.
2020-11-12 09:11:36,323 INFO    MainThread: DCP wide health check does not exist for amar
/Applications/Xcode.app/Contents/Developer/usr/bin/make grafana
python /Users/amar/dev/hca/azul/scripts/update_grafana.py
2020-11-12 09:11:36,655 INFO    MainThread: Skipping publishing of Grafana dashboard
/Applications/Xcode.app/Contents/Developer/usr/bin/make sam
python /Users/amar/dev/hca/azul/scripts/register_sam.py
2020-11-12 09:11:38,591 INFO    MainThread: Found credentials in environment variables.
Traceback (most recent call last):
  File "/Users/amar/dev/hca/azul/scripts/register_sam.py", line 37, in <module>
    main()
  File "/Users/amar/dev/hca/azul/scripts/register_sam.py", line 23, in main
    tdr.register_with_sam()
  File "/Users/amar/dev/hca/azul/src/azul/terra.py", line 153, in register_with_sam
    raise RequirementError('Unable to register %s. SAM does not allow '
azul.RequirementError: ('Unable to register %s. SAM does not allow re-registration of service account emails. Refer to the troubleshooting section of the README.', 'azul-ucsc-2-amar@human-cell-atlas-travis-test.iam.gserviceaccount.com')
make[2]: *** [sam] Error 1
make[1]: *** [apply] Error 2
make: *** [deploy] Error 2
```

**Review by amarjandu — CHANGES_REQUESTED** [review]
Very close! Please see the comment attached.

File: src/azul/terra.py [hunk 2]
```diff
@@ -150,9 +150,9 @@ def register_with_sam(self) -> None:
             log.info('Google service account previously registered with SAM.')
         elif (response.status == 500
               and 'Cannot update googleSubjectId' in response.data.decode()):
-            log.warning('Unable to register %s. SAM does not allow re-registration of '
-                        'service account emails. Refer to the troubleshooting section '
-                        'of the README.', self.credentials.service_account_email)
+            raise RequirementError(f'Unable to register {self.credentials.service_account_email}. '
```
Comment (amarjandu) [inline src/azul/terra.py:153]: trimmed line length cannot exceed 81 chars, this line is at 87. 
https://github.com/DataBiosphere/azul/blob/develop/CONTRIBUTING.rst#line-length

**Review by amarjandu — APPROVED** [review]
There is a line length issue, other than that LGTM.

**Review by hannes-ucsc — APPROVED** [review]
Could have fixed the line length issue that @amarjandu pointed out.

**Comment by hannes-ucsc** [comment]
For demo, run `make sam`, delete SA from TDR group tat grants access to the SA, run `make sam`, readd the SA, run `make sam`.

**Comment by noah-aviel-dove** [comment]
@hannes-ucsc: "[Noah] should come up with [their] own demo plan."

**Pull request merged**